pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/MinionInstancesCleanupTask.java

```diff
@@ -23,36 +23,61 @@
 
-  public MinionInstancesCleanupTask(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf controllerConf,
-      ControllerMetrics controllerMetrics) {
-    super("MinionInstancesCleanupTask", controllerConf.getMinionInstancesCleanupTaskFrequencyInSeconds(),
+  public MinionInstancesCleanupTask(PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerConf controllerConf, ControllerMetrics controllerMetrics) {
+    super(TASK_NAME, controllerConf.getMinionInstancesCleanupTaskFrequencyInSeconds(),
         controllerConf.getMinionInstancesCleanupTaskInitialDelaySeconds());
     _pinotHelixResourceManager = pinotHelixResourceManager;
+    _leadControllerManager = leadControllerManager;
     _controllerMetrics = controllerMetrics;
+    _minionInstanceCleanupTaskMaxOfflineTimeRangeInMilliseconds =
+        controllerConf.getMinionInstancesCleanupTaskMaxOfflineTimeRangeInSeconds() * 1000L;
   }
 
   @Override
   protected void runTask() {
+    // Make it so that only one controller returns the metric for all the tasks.
+    if (!_leadControllerManager.isLeaderForTable(TASK_NAME)) {
+      return;
+    }
+
     List<String> offlineInstances = new ArrayList<>(_pinotHelixResourceManager.getAllInstances());
     offlineInstances.removeAll(_pinotHelixResourceManager.getOnlineInstanceList());
     for (String offlineInstance : offlineInstances) {
+      // Since ZNodes under "/LIVEINSTANCES" are ephemeral, if there is a ZK session expire (e.g. due to network issue),
+      // the ZNode under "/LIVEINSTANCES" will be deleted. Thus, such race condition can happen when this task is running.
+      // In order to double confirm the live status of an instance, the field "LAST_OFFLINE_TIME" in ZNode under
+      // "/INSTANCES/<instance_id>/HISTORY" needs to be checked. If the value is "-1", that means the instance is ONLINE;
+      // if the value is a timestamp, that means the instance starts to be OFFLINE since that time.
       if (offlineInstance.startsWith(CommonConstants.Helix.PREFIX_OF_MINION_INSTANCE)) {
-        PinotResourceManagerResponse response = _pinotHelixResourceManager.dropInstance(offlineInstance);
-        if (response.isSuccessful()) {
-          _controllerMetrics.addValueToGlobalGauge(ControllerGauge.DROPPED_MINION_INSTANCES, 1);
+        // Drop the minion instance if it has been offline for more than a period of this task.
+        if (_pinotHelixResourceManager
+            .isInstanceOffline(offlineInstance, _minionInstanceCleanupTaskMaxOfflineTimeRangeInMilliseconds)) {
+          LOGGER.info("Dropping minion instance: {}", offlineInstance);
+          PinotResourceManagerResponse response = _pinotHelixResourceManager.dropInstance(offlineInstance);
+          if (response.isSuccessful()) {
+            _controllerMetrics.addValueToGlobalGauge(ControllerGauge.DROPPED_MINION_INSTANCES, 1);
```

Comment:
Maybe a better way is to add a minion guage that measures the number of online minions and emits it periodically in the  minion metrics task added earlier. This metric will appear as a blip and disapear.

Reply:
Sure, we can do that in a later PR.

Reply:
agreed